### PR TITLE
stream: support empty uv_try_write on unix

### PR DESCRIPTION
**Pull request by indutny**
cc @saghul

### CODE REVIEW

**Comment by bnoordhuis** [comment]
Can you add a small regression test?  Can probably be one or two lines added to an existing test.

**Comment by indutny** [comment]
@bnoordhuis added test, but it won't really show anything useful, because `uv_try_write` is not guaranteed to succeed.

File: test/test-tcp-try-write.c [hunk 1]
```diff
@@ -57,6 +57,10 @@ static void connect_cb(uv_connect_t* req, int status) {
       bytes_written += r;
       break;
     }
+
+    buf = uv_buf_init("", 0);
+    r = uv_try_write((uv_stream_t*) &client, &buf, 1);
+    ASSERT(r == 0 || r == UV_EAGAIN);
```
Comment (bnoordhuis) [inline test/test-tcp-try-write.c]: We can do a little better here.  If the previous uv_try_write() call failed with UV_EAGAIN, this one _must_ fail with UV_EAGAIN as well.

**Comment by bnoordhuis** [comment]
Can you confirm that no updates to src/win are necessary?

File: test/test-tcp-try-write.c (hunk 1 above)
Comment (bnoordhuis) [inline test/test-tcp-try-write.c]: Also, you should probably move this up a few lines, else it's not going to get reached when the previous uv_try_write() call succeeded.

Comment (indutny) [inline test/test-tcp-try-write.c]: Probably fixed.

**Comment by indutny** [comment]
I have checked the windows code. It looks OK.

**Comment by bnoordhuis** [comment]
LGTM

**Comment by indutny** [comment]
Landed in 5e761a5f812296aad8b44236f3363a9685c9c77e, thank you!

@saghul : when do you plan to do a next release? ;)

**Comment by saghul** [comment]
Sorry for coming late at this. Given that a read of length zero means that the connection was closed, what does it mean to try to send an empty packet? AFAIK it's not actually sent, is it?

IIRC this s discussed when we landed the PR which added support for empty UDP datagrams but it was left out for the reasons mentioned above.

**Comment by indutny** [comment]
@saghul I'm not sure if actually sends empty TCP packet, and I am too lame to check kernel implementation. Perhaps @bnoordhuis has insights?

**Comment by bnoordhuis** [comment]
Linux won't transmit a TCP packet for a zero-length write but I don't know about other operating systems.

**Pull request closed without merging**